File: R/R/outputs.R
# Copyright (c) Meta Platforms, Inc. and its affiliates.

# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

####################################################################
#' Evaluate Models and Output Results into Local Files
#'
#' Pack \code{robyn_plots()}, \code{robyn_csv()}, and \code{robyn_clusters()}
#' outcomes on \code{robyn_run()} results. When \code{UI=TRUE}, enriched
#' \code{OutputModels} results with additional plots and objects.
#'
#' @param InputCollect,OutputModels \code{robyn_inputs()} and \code{robyn_run()}
#' outcomes.
#' @param pareto_fronts Integer. Number of Pareto fronts for the output.
#' \code{pareto_fronts = 1} returns the best models trading off \code{NRMSE} &
#' \code{DECOMP.RSSD}. Increase \code{pareto_fronts} to get more model choices.
#' \code{pareto_fronts = "auto"} selects the min fronts that include at least 100
#' candidates. To customize this threshold, set value with \code{min_candidates}.
#' @param calibration_constraint Numeric. Default to 0.1 and allows 0.01-0.1. When
#' calibrating, 0.1 means top 10% calibrated models are used for pareto-optimal
#' selection. Lower \code{calibration_constraint} increases calibration accuracy.
#' @param plot_folder Character. Path for saving plots. Default
#' to \code{robyn_object} and saves plot in the same directory as \code{robyn_object}.
#' @param plot_folder_sub Character. Sub path for saving plots. Will overwrite the
#' default path with timestamp.
#' @param plot_pareto Boolean. Set to \code{FALSE} to deactivate plotting
#' and saving model one-pagers. Used when testing models.
#' @param clusters Boolean. Apply \code{robyn_clusters()} to output models?
#' @param select_model Character vector. Which models (by \code{solID}) do you
#' wish to plot the one-pagers and export? Default will take top
#' \code{robyn_clusters()} results.
#' @param csv_out Character. Accepts "pareto" or "all". Default to "pareto". Set
#' to "all" will output all iterations as csv. Set NULL to skip exports into CSVs.
#' @param ui Boolean. Save additional outputs for UI usage. List outcome.
#' @param export Boolean. Export outcomes into local files?
#' @param all_sol_json Logical. Add all solutions to json export.
#' @param quiet Boolean. Keep messages off?
#' @param refresh Boolean. Refresh mode
#' @param ... Additional parameters passed to \code{robyn_clusters()}
#' @return (Invisible) list. Class: \code{robyn_outputs}. Contains processed
#' results based on \code{robyn_run()} results.
#' @export
robyn_outputs <- function(InputCollect, OutputModels,
                          pareto_fronts = "auto",
                          calibration_constraint = 0.1,
                          plot_folder = NULL,
                          plot_folder_sub = NULL,
                          plot_pareto = TRUE,
                          csv_out = "pareto",
                          clusters = TRUE,
                          select_model = "clusters",
                          ui = FALSE, export = TRUE,
                          all_sol_json = FALSE,
                          quiet = FALSE,
                          refresh = FALSE, ...) {
  if (is.null(plot_folder)) plot_folder <- getwd()
  plot_folder <- check_dir(plot_folder)

  # Check calibration constrains
  calibrated <- !is.null(InputCollect$calibration_input)
  all_fixed <- length(OutputModels$trial1$hyperBoundFixed) == length(OutputModels$hyper_updated)
  if (!all_fixed) {
    calibration_constraint <- check_calibconstr(
      calibration_constraint,
      OutputModels$iterations,
      OutputModels$trials,
      InputCollect$calibration_input,
      refresh = refresh
    )
  }

  #####################################
  #### Run robyn_pareto on OutputModels

  totalModels <- OutputModels$iterations * OutputModels$trials
  if (!isTRUE(attr(OutputModels, "hyper_fixed"))) {
    message(sprintf(
      ">>> Running Pareto calculations for %s models on %s front%s...",
      totalModels, pareto_fronts, ifelse(pareto_fronts > 1, "s", "")
    ))
  }
  pareto_results <- robyn_pareto(
    InputCollect, OutputModels,
    pareto_fronts = pareto_fronts,
    calibration_constraint = calibration_constraint,
    quiet = quiet,
    calibrated = calibrated,
    ...
  )
  pareto_fronts <- pareto_results$pareto_fronts
  allSolutions <- pareto_results$pareto_solutions

  #####################################
  #### Gather the results into output object

  # Auxiliary list with all results (wasn't previously exported but needed for robyn_outputs())
  allPareto <- list(
    resultHypParam = pareto_results$resultHypParam,
    xDecompAgg = pareto_results$xDecompAgg,
    resultCalibration = pareto_results$resultCalibration,
    plotDataCollect = pareto_results$plotDataCollect,
    df_caov_pct = pareto_results$df_caov_pct_all
  )

  # Set folder to save outputs: legacy plot_folder_sub
  depth <- ifelse(
    "refreshDepth" %in% names(InputCollect),
    InputCollect$refreshDepth,
    ifelse("refreshCounter" %in% names(InputCollect),
      InputCollect$refreshCounter, 0
    )
  )
  folder_var <- ifelse(!as.integer(depth) > 0, "init", paste0("rf", depth))
  if (is.null(plot_folder_sub)) {
    plot_folder_sub <- paste("Robyn", format(Sys.time(), "%Y%m%d%H%M"), folder_var, sep = "_")
  }

  # Final results object
  OutputCollect <- list(
    resultHypParam = filter(pareto_results$resultHypParam, .data$solID %in% allSolutions),
    xDecompAgg = filter(pareto_results$xDecompAgg, .data$solID %in% allSolutions),
    mediaVecCollect = pareto_results$mediaVecCollect,
    xDecompVecCollect = pareto_results$xDecompVecCollect,
    resultCalibration = if (calibrated) {
      filter(pareto_results$resultCalibration, .data$solID %in% allSolutions)
    } else {
      NULL
    },
    allSolutions = allSolutions,
    allPareto = allPareto,
    calibration_constraint = calibration_constraint,
    OutputModels = OutputModels,
    cores = OutputModels$cores,
    iterations = OutputModels$iterations,
    trials = OutputModels$trials,
    intercept_sign = OutputModels$intercept_sign,
    nevergrad_algo = OutputModels$nevergrad_algo,
    add_penalty_factor = OutputModels$add_penalty_factor,
    seed = OutputModels$seed,
    UI = NULL,
    pareto_fronts = pareto_fronts,
    hyper_fixed = attr(OutputModels, "hyper_fixed"),
    plot_folder = gsub("//", "", paste0(plot_folder, "/", plot_folder_sub, "/"))
  )
  class(OutputCollect) <- c("robyn_outputs", class(OutputCollect))

  plotPath <- paste0(plot_folder, "/", plot_folder_sub, "/")
  OutputCollect$plot_folder <- gsub("//", "/", plotPath)
  if (export && !dir.exists(OutputCollect$plot_folder)) dir.create(OutputCollect$plot_folder, recursive = TRUE)

  # Cluster results and amend cluster output
  if (clusters) {
    if (!quiet) message(">>> Calculating clusters for model selection using Pareto fronts...")
    try(clusterCollect <- robyn_clusters(OutputCollect,
      dep_var_type = InputCollect$dep_var_type,
      quiet = quiet, export = export, ...
    ))
    OutputCollect$resultHypParam <- left_join(
      OutputCollect$resultHypParam,
      select(clusterCollect$data, .data$solID, .data$cluster, .data$top_sol),
      by = "solID"
    )
    OutputCollect$xDecompAgg <- left_join(
      OutputCollect$xDecompAgg,
      select(clusterCollect$data, .data$solID, .data$cluster, .data$top_sol),
      by = "solID"
    ) %>%
      left_join(
        select(
          clusterCollect$df_cluster_ci, .data$rn, .data$cluster, .data$boot_mean,
          .data$boot_se, .data$ci_low, .data$ci_up, .data$rn
        ),
        by = c("rn", "cluster")
      ) %>%
      left_join(
        pareto_results$df_caov_pct_all,
        by = c("solID", "rn")
      )
    OutputCollect$mediaVecCollect <- left_join(
      OutputCollect$mediaVecCollect,
      select(clusterCollect$data, .data$solID, .data$cluster, .data$top_sol),
      by = "solID"
    )
    OutputCollect$xDecompVecCollect <- left_join(
      OutputCollect$xDecompVecCollect,
      select(clusterCollect$data, .data$solID, .data$cluster, .data$top_sol),
      by = "solID"
    )
    if (calibrated) {
      OutputCollect$resultCalibration <- left_join(
        OutputCollect$resultCalibration,
        select(clusterCollect$data, .data$solID, .data$cluster, .data$top_sol),
        by = "solID"
      )
    }
    OutputCollect[["clusters"]] <- clusterCollect
  }

  if (export) {
    tryCatch(
      {
        if (!quiet) message(paste0(">>> Collecting ", length(allSolutions), " pareto-optimum results into: ", OutputCollect$plot_folder))

        if (!quiet) message(">> Exporting general plots into directory...")
        all_plots <- robyn_plots(InputCollect, OutputCollect, export = export)

        if (csv_out %in% c("all", "pareto")) {
          if (!quiet) message(paste(">> Exporting", csv_out, "results as CSVs into directory..."))
          robyn_csv(InputCollect, OutputCollect, csv_out, export = export, calibrated = calibrated)
        }

        if (plot_pareto) {
          if (!quiet) {
            message(sprintf(
              ">>> Exporting %sone-pagers into directory...", ifelse(!OutputCollect$hyper_fixed, "pareto ", "")
            ))
          }
          select_model <- if (!clusters || is.null(OutputCollect[["clusters"]])) NULL else select_model
          pareto_onepagers <- robyn_onepagers(
            InputCollect, OutputCollect,
            select_model = select_model,
            quiet = quiet,
            export = export
          )
        }

        if (all_sol_json) {
          all_sol_json <- OutputCollect$resultHypParam %>%
            filter(!is.na(.data$cluster)) %>%
            select(c("solID", "cluster", "top_sol")) %>%
            arrange(.data$cluster, -.data$top_sol, .data$solID)
        } else {
          all_sol_json <- NULL
        }
        robyn_write(
          InputCollect = InputCollect, OutputModels = OutputModels,
          dir = OutputCollect$plot_folder, quiet = quiet, all_sol_json = all_sol_json
        )

        # For internal use -> UI Code
        if (ui && plot_pareto) OutputCollect$UI$pareto_onepagers <- pareto_onepagers
        OutputCollect[["UI"]] <- if (ui) list(pParFront = all_plots[["pParFront"]]) else NULL
      },
      error = function(err) {
        message(paste("Failed exporting results, but returned model results anyways:\n", err))
      }
    )
  }

  if (!is.null(OutputModels$hyper_updated)) OutputCollect$hyper_updated <- OutputModels$hyper_updated
  class(OutputCollect) <- c("robyn_outputs", class(OutputCollect))
  return(invisible(OutputCollect))
}

#' @rdname robyn_outputs
#' @aliases robyn_outputs
#' @param x \code{robyn_outputs()} output.
#' @export
print.robyn_outputs <- function(x, ...) {
  print(glued(
    "
Plot Folder: {x$plot_folder}
Calibration Constraint: {x$calibration_constraint}
Hyper-parameters fixed: {x$hyper_fixed}
Pareto-front ({x$pareto_fronts}) All solutions ({nSols}): {paste(x$allSolutions, collapse = ', ')}
{clusters_info}
",
    nSols = length(x$allSolutions),
    clusters_info = if ("clusters" %in% names(x)) {
      glued(
        "Clusters (k = {x$clusters$n_clusters}): {paste(x$clusters$models$solID, collapse = ', ')}"
      )
    } else {
      NULL
    }
  ))
}


####################################################################
#' Output results into local files: CSV files
#'
#' @param OutputCollect \code{robyn_run(..., export = FALSE)} output.
#' @param calibrated Logical
#' @rdname robyn_outputs
#' @return Invisible \code{NULL}.
#' @export
robyn_csv <- function(InputCollect, OutputCollect, csv_out = NULL, export = TRUE, calibrated = FALSE) {
  if (export) {
    check_class("robyn_outputs", OutputCollect)
    temp_all <- OutputCollect$allPareto
    if ("pareto" %in% csv_out) {
      write.csv(OutputCollect$resultHypParam, paste0(OutputCollect$plot_folder, "pareto_hyperparameters.csv"))
      write.csv(OutputCollect$xDecompAgg, paste0(OutputCollect$plot_folder, "pareto_aggregated.csv"))
      if (calibrated) {
        write.csv(OutputCollect$resultCalibration, paste0(OutputCollect$plot_folder, "pareto_calibration.csv"))
      }
    }
    if ("all" %in% csv_out) {
      write.csv(temp_all$resultHypParam, paste0(OutputCollect$plot_folder, "all_hyperparameters.csv"))
      write.csv(temp_all$xDecompAgg, paste0(OutputCollect$plot_folder, "all_aggregated.csv"))
      if (calibrated) {
        write.csv(temp_all$resultCalibration, paste0(OutputCollect$plot_folder, "all_calibration.csv"))
      }
    }
    if (!is.null(csv_out)) {
      write.csv(InputCollect$dt_input, paste0(OutputCollect$plot_folder, "raw_data.csv"))
      write.csv(OutputCollect$mediaVecCollect, paste0(OutputCollect$plot_folder, "pareto_media_transform_matrix.csv"))
      write.csv(OutputCollect$xDecompVecCollect, paste0(OutputCollect$plot_folder, "pareto_alldecomp_matrix.csv"))
    }
  }
}

File: python/src/robyn/outputs.py
# Copyright (c) Meta Platforms, Inc. and its affiliates.

# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

####################################################################
import os
import time
import pandas as pd
import re
from datetime import datetime

from .json import robyn_write
from .cluster import robyn_clusters
from .checks import check_dir, check_calibconstr
from .pareto import robyn_pareto

def robyn_outputs(input_collect,
                  output_models,
                  pareto_fronts="auto",
                  calibration_constraint=0.1,
                  plot_folder=None,
                  plot_folder_sub=None,
                  plot_pareto=True,
                  csv_out="pareto",
                  clusters=True,
                  select_model="clusters",
                  ui=False,
                  export=True,
                  all_sol_json=False,
                  quiet=False,
                  refresh=False):
    """
    Runs the Robyn Pareto algorithm on the given output models and collects the results.

    Args:
        input_collect (dict): The input collection.
        output_models (object): The output models.
        pareto_fronts (str, optional): The number of Pareto fronts to calculate. Defaults to "auto".
        calibration_constraint (float, optional): The calibration constraint. Defaults to 0.1.
        plot_folder (str, optional): The folder to save the plots. Defaults to None.
        plot_folder_sub (str, optional): The subfolder within the plot folder. Defaults to None.
        plot_pareto (bool, optional): Whether to plot the Pareto fronts. Defaults to True.
        csv_out (str, optional): The type of CSV output. Defaults to "pareto".
        clusters (bool, optional): Whether to calculate clusters for model selection. Defaults to True.
        select_model (str, optional): The model selection method. Defaults to "clusters".
        ui (bool, optional): Whether to enable the user interface. Defaults to False.
        export (bool, optional): Whether to export the results. Defaults to True.
        all_sol_json (bool, optional): Whether to export all solutions as JSON. Defaults to False.
        quiet (bool, optional): Whether to suppress console output. Defaults to False.
        refresh (bool, optional): Whether to refresh the results. Defaults to False.

    Returns:
        dict: The collected output results.
    """

    t0 = time.time()

    if plot_folder is None:
        plot_folder = os.getcwd()

    plot_folder = check_dir(plot_folder)

    # Check calibration constrains
    calibrated = 'calibration_input' in input_collect and input_collect['calibration_input'] is not None
    all_fixed = len(output_models["trials"][0]["hyperBoundFixed"]) == len(output_models["hyper_updated"])
    if not all_fixed:
        calibration_constraint = check_calibconstr(calibration_constraint, output_models["metadata"]["iterations"], output_models["metadata"]["trials"], input_collect["robyn_inputs"]["calibration_input"], refresh=refresh)

    #####################################
    #### Run robyn_pareto on OutputModels

    total_models = output_models["metadata"]["iterations"] * output_models["metadata"]["trials"]
    if not isinstance(output_models["metadata"]["hyper_fixed"], bool):
        print(f"Running Pareto calculations for {total_models} models on {pareto_fronts} fronts...")

    pareto_results = robyn_pareto(input_collect, output_models, pareto_fronts=pareto_fronts, calibration_constraint=calibration_constraint, quiet=quiet, calibrated=calibrated, refresh=refresh)
    pareto_fronts = pareto_results["pareto_fronts"]
    all_solutions = pareto_results["pareto_solutions"]

    #####################################
    #### Gather the results into output object

    all_pareto = {
        "resultHypParam": pareto_results["resultHypParam"],
        "xDecompAgg": pareto_results["xDecompAgg"],
        "resultCalibration": pareto_results["resultCalibration"],
        "plotDataCollect": pareto_results["plotDataCollect"],
        "df_caov_pct": pareto_results["df_caov_pct_all"]
    }

    # Set folder to save outputs
    depth = 0 if "refreshDepth" not in input_collect["robyn_inputs"] else input_collect["robyn_inputs"]["refreshDepth"]
    folder_var = "init" if not int(depth) > 0 else "rf" + str(depth)

    if plot_folder_sub is None:
        plot_folder_sub = "Robyn_" + datetime.now().strftime("%Y%m%d%H%M") + "_" + folder_var

    plot_folder = re.sub("//+", "/", f"{plot_folder}/{plot_folder_sub}/")

    if not os.path.exists(plot_folder) and export:
        print(f"Creating directory for outputs: {plot_folder}")
        os.makedirs(plot_folder)

    # Final results object
    OutputCollect = {
        'resultHypParam': pareto_results['resultHypParam'][pareto_results['resultHypParam']['solID'].isin(all_solutions)],
        'xDecompAgg': pareto_results['xDecompAgg'][pareto_results['xDecompAgg']['solID'].isin(all_solutions)],

        #"xDecompAgg": pareto_results["xDecompAgg"].loc[pareto_results["solID"].isin(allSolutions)],
        "mediaVecCollect": pareto_results["mediaVecCollect"],
        "xDecompVecCollect": pareto_results["xDecompVecCollect"],
        #"resultCalibration": None if not calibrated else pareto_results["resultCalibration"].loc[pareto_results["solID"].isin(allSolutions)],
        "resultCalibration": pareto_results["resultCalibration"][pareto_results["resultCalibration"]["solID"].isin(all_solutions)] if calibrated else None,
        "allSolutions": all_solutions,
        "allPareto": all_pareto,
        "calibration_constraint": calibration_constraint,
        "OutputModels": output_models,
        "cores": output_models["metadata"]["cores"],
        "iterations": output_models["metadata"]["iterations"],
        "trials": output_models["trials"],
        "intercept_sign": output_models["metadata"]["intercept_sign"],
        "nevergrad_algo": output_models["metadata"]["nevergrad_algo"],
        "add_penalty_factor": output_models["metadata"]["add_penalty_factor"],
        "seed": output_models["seed"],
        "UI": None,
        "pareto_fronts": pareto_fronts,
        'hyper_fixed': output_models["metadata"]['hyper_fixed'],
        "plot_folder": plot_folder
    }
    OutputCollect.keys()

    # Cluster results and amend cluster output
    if clusters:
        if not quiet:
            print(">>> Calculating clusters for model selection using Pareto fronts...")
        clusterCollect = robyn_clusters(
            OutputCollect,
            dep_var_type=input_collect["robyn_inputs"]["dep_var_type"],
            quiet=quiet,
            export=export
        )
        OutputCollect["resultHypParam"] = pd.merge(
            OutputCollect["resultHypParam"],
            clusterCollect["data"].loc[clusterCollect["data"]["solID"].isin(all_solutions)],
            on="solID"
        )
        OutputCollect["xDecompAgg"] = pd.merge(
            OutputCollect["xDecompAgg"],
            clusterCollect["data"].loc[clusterCollect["data"]["solID"].isin(all_solutions)],
            on="solID"
        )
        OutputCollect["mediaVecCollect"] = pd.merge(
            OutputCollect["mediaVecCollect"],
            clusterCollect["data"].loc[clusterCollect["data"]["solID"].isin(all_solutions)],
            on="solID"
        )
        OutputCollect["xDecompVecCollect"] = pd.merge(
            OutputCollect["xDecompVecCollect"],
            clusterCollect["data"].loc[clusterCollect["data"]["solID"].isin(all_solutions)],
            on="solID"
        )
        if calibrated:
            OutputCollect["resultCalibration"] = pd.merge(
                OutputCollect["resultCalibration"],
                clusterCollect["data"].loc[clusterCollect["data"]["solID"].isin(all_solutions)],
                on="solID"
            )
        OutputCollect["clusters"] = clusterCollect

    # TODO Add export code to enable plotting
    # if export:
    #     try:
    #         message(">>> Collecting {} pareto-optimum results into: {}".format(len(all_solutions), plot_folder))
    #         all_plots = robyn_plots(input_collect, output_collect, export=export, quiet=quiet)
    #         message(">> Exporting general plots into directory...")
    #         if csv_out in ["all", "pareto"]:
    #             message(">> Exporting {} results as CSVs into directory...".format(csv_out))
    #             robyn_csv(input_collect, output_collect, csv_out, export=export, calibrated=calibrated)
    #         if plot_pareto:
    #             message(">>> Exporting pareto one-pagers into directory...")
    #             select_model = select_model if not clusters or output_collect["clusters"] is None else None
    #             pareto_onepagers = robyn_onepagers(input_collect, output_collect, select_model=select_model, quiet=quiet, export=export)
    #         if all_sol_json:
    #             pareto_df = output_collect["resultHypParam"].filter(pandas.notnull(pandas.Series(["cluster"]))).select(["solID", "cluster", "top_sol"]).sort_values(by=["cluster", "top_sol"], ascending=False).drop(columns=["solID"])
    #         else:
    #             pareto_df = None
    #         ##attr(output_collect, "runTime") = round(difftime(sys.time(), t0, units="mins"), 2)
    #         output_collect["runTime"] = round(difftime(sys.time(), t0, units="mins"), 2)
    #         robyn_write(input_collect, output_collect, dir=plot_folder, quiet=quiet, pareto_df=pareto_df, export=export)
    #         if ui and plot_pareto:
    #             output_collect["UI"] = {"pareto_onepagers": pareto_onepagers}
    #         output_collect["UI"] = output_collect.get("UI", pandas.DataFrame()) if ui else None
    #     except Exception as e:
    #         message("Failed exporting results, but returned model results anyways: {}".format(e))
    ##if not is.null(output_models["hyper_updated"]):
    if output_models["hyper_updated"] is not None:
        OutputCollect["hyper_updated"] = output_models["hyper_updated"]
    ##attr(output_collect, "runTime") = round(difftime(sys.time(), t0, units="mins"), 2)
    # OutputCollect["runTime"] = round(difftime(sys.time(), t0, units="mins"), 2)
    ##class(output_collect) = ["robyn_outputs", class(output_collect)]
    ##??output_collect["robyn_outputs"] = output_collect
    ##return(invisible(output_collect))
    return OutputCollect

def print_robyn_outputs(x, *args, **kwargs):
    """
    Print various outputs related to Robyn.

    Parameters:
    - x: Robyn object
    - *args: Additional positional arguments
    - **kwargs: Additional keyword arguments
    """
    print("Plot Folder: {x.plot_folder}")
    print("Calibration Constraint: {x.calibration_constraint}")
    print("Hyper-parameters fixed: {x.hyper_fixed}")
    print("Pareto-front ({x.pareto_fronts}) All solutions ({len(x.allSolutions)}): {', '.join(x.allSolutions)}")
    if "clusters" in x.keys():
        print("Clusters (k = {x.clusters.n_clusters}): {', '.join(x.clusters.models.solID)}")
    else:
        print("")

def robyn_csv(input_collect, output_collect, csv_out=None, export=True, calibrated=False):
    """
    Export data from Robyn outputs to CSV files.

    Args:
        input_collect (InputCollect): The input collection object.
        output_collect (robyn_outputs): The output collection object.
        csv_out (str or None, optional): The type of CSV files to export. Defaults to None.
        export (bool, optional): Whether to export the data to CSV files. Defaults to True.
        calibrated (bool, optional): Whether the data is calibrated. Defaults to False.
    """
    if export:
        # Check that OutputCollect has the correct class
        assert isinstance(output_collect, robyn_outputs)

        # Get the temp all dataframe
        temp_all = output_collect.allPareto

        # Get the plot folder
        plot_folder = output_collect.plot_folder

        # Write the pareto hyperparameters and aggregated data to CSV
        if "pareto" in csv_out:
            pd.write_csv(output_collect.resultHypParam, os.path.join(plot_folder, "pareto_hyperparameters.csv"))
            pd.write_csv(output_collect.xDecompAgg, os.path.join(plot_folder, "pareto_aggregated.csv"))
            if calibrated:
                pd.write_csv(output_collect.resultCalibration, os.path.join(plot_folder, "pareto_calibration.csv"))

        # Write the all hyperparameters and aggregated data to CSV
        if "all" in csv_out:
            pd.write_csv(temp_all.resultHypParam, os.path.join(plot_folder, "all_hyperparameters.csv"))
            pd.write_csv(temp_all.xDecompAgg, os.path.join(plot_folder, "all_aggregated.csv"))
            if calibrated:
                pd.write_csv(temp_all.resultCalibration, os.path.join(plot_folder, "all_calibration.csv"))

        # Write the raw data and transformation matrices to CSV
        ## if not is.null(csv_out):
        if csv_out is not None:
            pd.write_csv(input_collect.dt_input, os.path.join(plot_folder, "raw_data.csv"))
            pd.write_csv(output_collect.mediaVecCollect, os.path.join(plot_folder, "pareto_media_transform_matrix.csv"))
            pd.write_csv(output_collect.xDecompVecCollect, os.path.join(plot_folder, "pareto_alldecomp_matrix.csv"))
